### ISSUE TITLE
Recurring job that runs on a non default queue is retried on default

### ISSUE THREAD

**Issue by tstojecki** [HangfireIO/Hangfire#705]
When a recurring job that is scheduled to run on a non default fails it is retried on the default one:

The job is scheduled this way:
`app.UseHangfireServer(new BackgroundJobServerOptions { Queues = new[] { "WEB" } })`

In hangfire console, the job shows the following:

Enqueued
Triggered by DelayedJobScheduler
Queue:
**DEFAULT**

Failed
An exception occurred during performance of the job.

Enqueued
Triggered by recurring job scheduler
Queue:
**WEB**

Is this a bug?

**Comment by pbolduc** [comment]
I have seen this problem as well.  I have one set of HangFire tables in my database supported two related applications.  One app uses queue **A** and **default**.  The other app uses queue **B**.  Occasionally, a job in queue **B** will fail and it is rescheduled on the **default** queue.  The retried job always fails again because the first app does not have permissions/configuration to support the failed job.

I am running: 1.6.6

**Comment by pbolduc** [comment]
When [DelayedJobSchedule enqueues](https://github.com/HangfireIO/Hangfire/blob/129707d66fde24dc6379fb9d6b15fa0b8ca48605/src/Hangfire.Core/Server/DelayedJobScheduler.cs#L160-L165) the job, it does not set the queue in the EnqueuedState. The [default constructor for EnqueuedState](https://github.com/HangfireIO/Hangfire/blob/129707d66fde24dc6379fb9d6b15fa0b8ca48605/src/Hangfire.Core/States/EnqueuedState.cs#L88-L91) sets the queue to DefaultQueue (default). 

In my database, all of my jobs are in state Enqueued, however, in the JobQueue table, the queue is default.  I am adding my job using this code:

```
RecurringJob.AddOrUpdate<IJobType>(name, job => job.Run(), Cron.Hourly(), queue: "application");
```

**Comment by mikeholczer** [comment]
I don't think this is limited to recurring jobs. I'm using 1.6.8 and seeing the same thing for one-time jobs I created with BackgroundJobClient.Create and specify a non-default queue.

**Comment by mikeholczer** [comment]
@pbolduc looks like you diagnosed the issue pretty specifically, have you spent any time trying to fix it. I was thinking about giving it a stab.

**Comment by pbolduc** [comment]
I tried look at it, however, where it gets re-queued, the original queue is not available.  I looked into how I could get the original queue name, but could not figure it out in the time I had.  Eventually, I just created a separate database for the 2nd application. :-(

**Comment by kbrimington** [comment]
I panicked a bit when reading over this issue. We're transitioning to a microservices approach, and were depending on queues to provide separated concerns between apps while sharing a single Hangfire.

I figured out a workaround for you to consider. Here is a sample job. Note the QueueAttribute on the Run method.

    public class CheeseJob
    {
        [Queue("root")]
        public void Run()
        {
            throw new System.Exception("Things are not happy");
        }
    }

    // ...
    RecurringJob.AddOrUpdate<CheeseJob>("Cheese", job => job.Run(), Cron.Hourly(), queue: "root");

After giving this a run and letting it retry a few times, I queried the Job table and got this:

![image](https://cloud.githubusercontent.com/assets/21696007/22767245/5b20b28c-ee37-11e6-85bc-98bf96efc7bb.png)

I gave it another try, this time removing the QueueAttribute from the job method. This produced the queue-hopping symptom.

    public class CheeseJob
    {
        public void Run()
        {
            throw new System.Exception("Things are not happy");
        }
    }

    // ...
    RecurringJob.AddOrUpdate<CheeseJob>("Cheese", job => job.Run(), Cron.Hourly(), queue: "root");

![image](https://cloud.githubusercontent.com/assets/21696007/22767457/8cb9ce9a-ee38-11e6-8f66-4a2182361ca2.png)

From this information, it would seem that you can work around the queue hopping issue by applying a QueueAttribute to the job method.

Even so, I would really like to see a fix for the queue hopping behavior for two reasons: the behavior is unexpected, and it is prohibitive to folk who have some reason not to perma-assign the job to a single queue using an attribute.

**Comment by brunopessanha** [comment]
Requeue to initial queue #502

**Comment by brunopessanha** [comment]
There's an attribute in **Requeue to initial queue #502** that also changes the behaviour to enforce the original queue rather than the default one.
